enhancements/baremetal/baremetal-config-raid-and-bios.md

```diff
@@ -0,0 +1,149 @@
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes to add the support of RAID and BIOS configuration for baremetal IPI deployments.
+
+## Motivation
+
+With the new features of [RAID configuration](https://github.com/metal3-io/baremetal-operator/pull/292) and [BIOS configuration](https://github.com/metal3-io/baremetal-operator/pull/302), openshift already has the ability to configure RAID and BIOS for baremetal.
+
+However, these features only work with a running cluster, users cannot configure desired RAID and BIOS during IPI deployments.
+
+### Goals
+
+- Allow users to manage the RAID and BIOS configuration for control plane hosts during installation.
+- IPI deployments can handle the configuration of RAID and BIOS.
+
+### Non-Goals
+
+- Add new or vendor specific options regarding RAID and BIOS.
+
+## Proposal
+
+The configuration of RAID and BIOS is finally handed over to Ironic during the IPI deployments.
+1. The user runs `create manifests` and then adds RAID and BIOS fields to the **BMH** created by the installer.
+2. The interface to terraform in the installer code parses the BIOS and RAID fields.
+3. [Terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic) gets the
+configuration of RAID and BIOS for control plane hosts and then passes it to Ironic([Baremetal Operator(BMO)](https://github.com/metal3-io/baremetal-operator)
+has supported the process of RAID and BIOS for worker nodes).
+
+### User Stories
+
+With the addition of this feature, the users can configure RAID and BIOS in IPI deployments.
+
+### Implementation Details/Notes/Constraints
+
+#### Fill two fields
+
+The user runs `create manifests` and then fill *raid* and *firmware* fields in the **BMH** created by the installer.
+
+```yaml
+spec:
+  bmc:
+    address: <bmc>://<ip-address>
+    credentialsName: worker-0-bmc-secret
+  bootMACAddress: mac-address
+  raid: raid-config
+  firmware: bios-config
+  ...
+```
+
+#### Process the fields in installer
+
+For control plane hosts, transform the *raid* and *firmware* fields into string respectively and then write them into the
+***terraform.baremetal.auto.tfvars.json*** file as the values of *raid_config* and *bios_settings* respectively.
+
+#### Process the fields in terraform-provider-ironic
+
+Add *raid_config* and *bios_settings* fields to terraform-provider-ironic API, transform them back to struct format, and then call the
+methods in **BMO** to process the two fields to perform [manual cleaning](https://docs.openstack.org/ironic/latest/admin/cleaning.html#manual-cleaning).
```

Comment:
This sounds like we'll make terraform-provider-ironic run BMO code, but IIUC that's not the case - we'll implement equivalent functionality in terraform-provider-ironic via gophercloud directly, right?

Reply:
Sorry, currently we did use some methods in BMO. If needed, we can implement equivalent functionality in terraform-provider-ironic via gophercloud directly.

Reply:
Hi! Ping! PTAL @hardys

Reply:
I think implementing the logic in the terraform-provider-ironic is the way to go. The code there doesn't vendor BMO and we'd like to keep it that way.

Reply:
@kirankt Thanks for your review, but the community has [discussed this matter](https://github.com/openshift-metal3/terraform-provider-ironic/pull/53#issuecomment-917993321), and the final decision is to use BMO's existing code, but the BMO code structure will be adjusted, and [the adjustment is still under discussion](https://github.com/metal3-io/baremetal-operator/pull/1017).

Reply:
From what I read, I don't think there is consensus there yet. @stbenjam what are you thoughts on vendoring BMO into the ironic TF provider.

Reply:
@kirankt I updated the section `Process the fields in terraform-provider-ironic`, PTAL.

Reply:
@kirankt, I got a comment from @zaneb under PR53, and he suggested [import all of the bmo modules](https://github.com/openshift-metal3/terraform-provider-ironic/pull/53#discussion_r762031284) on terraform-provider-ironic side. I also think that it is fine to vendor bmo modules because it can avoid reinventing the wheel. WDYT?

Reply:
We [are already](https://github.com/openshift/installer/blob/master/go.mod#L480) in the installer so it seems to me it can't make anything worse.

Reply:
I reverted the section `Process the fields in terraform-provider-ironic` accordingly, PTAL.